src/main/java/com/ita/if103java/ims/service/impl/WarehouseServiceImpl.java

```diff
@@ -44,44 +64,121 @@ public WarehouseDto add(WarehouseDto warehouseDto) {
         Integer maxWarehouseDepth = userDetails.getAccountType().getMaxWarehouseDepth();
         int parentLevel = warehouseDao.findLevelByParentID(warehouseDto.getParentID());
         if (parentLevel + 1 < maxWarehouseDepth) {
-            return createNewWarehouse(warehouseDto);
+            return createNewWarehouse(warehouseDto, userDetails);
         } else {
             throw new MaxWarehouseDepthLimitReachedException("The maximum depth of warehouse's levels has been reached for this" +
                 "{accountId = " + accountId + "}");
         }
     }
 
-    private WarehouseDto createNewWarehouse(WarehouseDto warehouseDto) {
-        Warehouse warehouse = warehouseDtoMapper.toEntity(warehouseDto);
-        return warehouseDtoMapper.toDto(warehouseDao.create(warehouse));
+    private WarehouseDto createNewWarehouse(WarehouseDto warehouseDto, UserDetailsImpl user) {
+        Warehouse warehouse = warehouseDao.create(warehouseDtoMapper.toEntity(warehouseDto));
+        Address address = addressDtoMapper.toEntity(warehouseDto.getWarehouseAddressDto());
+        addressDao.createWarehouseAddress(warehouse.getId(), address);
+        createEvent(user, warehouse, EventName.WAREHOUSE_CREATED);
+
+        Warehouse newWarehouse = warehouseDao.create(warehouse);
+        populatePath(newWarehouse, user);
+        return warehouseDtoMapper.toDto(newWarehouse);
```

Comment:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! For Next PR !!!!!!!!!!!!!!
Need to create address only for top level warehouse.
Need to return address to UI in all find* methods if warehouse is the top level